File: src/net/acmicpc/solved/gold/Solution_1976.py
import sys

# libs
input = sys.stdin.readline


# 여행 가자 - G4
# https://www.acmicpc.net/problem/1976


def find(parent, x):
    # find 연산
    if parent[x] != x:
        parent[x] = find(parent, parent[x])

    return parent[x]


def union(parent, a, b):
    # union 연산
    root_a = find(parent, a)
    root_b = find(parent, b)

    if root_a != root_b:
        parent[root_b] = root_a


# 1. 입력 받기
N = int(input())  # N - 도시의 수
M = int(input())  # M - 여행 계획에 포함된 도시의 수

# Union-Find 배열 초기화
parent = list(range(N + 1))

# 2. 도시 연결 정보 입력 및 도시 연결 (union 연산)
for i in range(1, N + 1):
    connections = list(map(int, input().split()))
    for j in range(N):
        if connections[j] == 1:
            union(parent, i, j + 1)  # i번 도시와 (j + 1)번 도시를 연결

# 3. 여행 계획 입력
travel_plan = list(map(int, input().split()))

# 4. 여행 계획의 모든 도시가 같은 그룹에 속하는지 확인 (find 연산)
first_city_group = find(parent, travel_plan[0])
possible = True
for city in travel_plan[1:]:
    if find(parent, city) != first_city_group:
        possible = False
        break

# 5. 결과 출력
if possible:
    print("YES")
else:
    print("NO")

File: src/net/acmicpc/solved/gold/Solution_1976.java
package net.acmicpc.solved.gold;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

/*
 * 여행 가자 - G4
 * https://www.acmicpc.net/problem/1976
 */
public class Solution_1976 {

	static int[] parents;

	public static void main(String[] args) throws IOException {
		int N = nextInt(), M = nextInt();

		// 무리 초기화
		parents = new int[N + 1];
		for (int i = 1; i <= N; i++) {
			parents[i] = i;
		}

		// 여행 계획 입력
		for (int i = 1; i <= N; i++) {
			for (int j = 1; j <= N; j++) {
				int val = nextInt();
				if (val == 1) {
					union(i, j);
				}
			}
		}

		// 여행 계획 수행
		boolean flag = true;
		int root = -1;
		for (int i = 0; i < M; i++) {
			if (root == -1) {
				root = find(nextInt());
				continue;
			}

			// 하나라도 조상이 다른 경우, 실패
			if (find(nextInt()) != root) {
				flag = false;
				break;
			}
		}


		System.out.print(flag ? "YES" : "NO");
		br.close();
	}

	static int find(int x) {
		if (parents[x] == x) {
			return x;
		}

		return parents[x] = find(parents[x]);
	}

	static boolean union(int x, int y) {
		int rx = find(x);
		int ry = find(y);

		if (rx != ry) {
			parents[ry] = rx;
			return true;
		}

		return false;
	}


	// Libraries
	static String nextToken() throws IOException {
		while (st == null || !st.hasMoreTokens()) {
			String s = br.readLine();
			if (s == null) {
				return null;
			}
			st = new StringTokenizer(s);
		}
		return st.nextToken();
	}

	static int nextInt() throws IOException {
		return Integer.parseInt(nextToken());
	}

	static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	static StringBuilder sb = new StringBuilder();
	static StringTokenizer st;

}
